### ISSUE TITLE
Oracle array of object containing date mapped to LocalDate cannot be bound as a variable

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6967]
Consider this script:

```sql
CREATE TYPE u_2155_array AS VARRAY(10) OF DATE
/

CREATE TYPE u_2155_object AS OBJECT (
  D DATE,
  A u_2155_array
)
/

CREATE PROCEDURE p2155_1 (o OUT u_2155_object) IS
BEGIN
  o := u_2155_object(DATE '2017-01-01', u_2155_array(DATE '2017-01-01', DATE '2017-01-01'));
END p2155_1;
/

CREATE PROCEDURE p2155_2 (i u_2155_object, o OUT u_2155_object) IS
BEGIN
  o := i;
END p2155_2;
/
```

When activating `<javaTimeTypes/>` and running the following test:

```java
@Test
public void testOracleJSR310TypesOnUDTs() {
    LocalDate d = LocalDate.parse("2017-01-01");

    org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ArrayRecord a =
        new org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ArrayRecord(d, d);

    org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ObjectRecord o1 =
        new org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ObjectRecord(d, a);

    org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ObjectRecord o2 =
        org.jooq.test.oracle310.generatedclasses.Routines.p2155_1(configuration());

    assertEquals(o2, o1);

    org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ObjectRecord o3 =
        org.jooq.test.oracle310.generatedclasses.Routines.p2155_2(configuration(), o1);

    assertEquals(o3, o1);

    Record1<org.jooq.test.oracle310.generatedclasses.udt.records.U_2155ObjectRecord> r =
    create().select(val(o1))
            .fetchOne();

    assertEquals(d, r.value1().getD());
    assertEquals(d, r.value1().getA().get(0));
    assertEquals(d, r.value1().getA().get(1));
}
```

The first test passes in jOOQ 3.11 (but not in jOOQ 3.10), and the second and third tests fail with the following exception:

```
org.jooq.exception.DataAccessException: SQL [null]; Error while writing value at JDBC bind index: 1
	at org.jooq_3.11.0.ORACLE.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2238)
	at org.jooq.impl.AbstractBindContext.bindValue(AbstractBindContext.java:127)
	at org.jooq.impl.UDTConstant.bind0(UDTConstant.java:175)
	at org.jooq.impl.UDTConstant.accept(UDTConstant.java:68)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:254)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:427)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:188)
	at org.jooq.impl.AbstractRoutine.bind1(AbstractRoutine.java:626)
	at org.jooq.impl.AbstractRoutine.bind0(AbstractRoutine.java:588)
	at org.jooq.impl.AbstractRoutine.accept(AbstractRoutine.java:531)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:254)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:427)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:188)
	at org.jooq.impl.AbstractRoutine.executeCallableStatement(AbstractRoutine.java:468)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:339)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:316)
	at org.jooq.test.oracle310.generatedclasses.Routines.p2155_2(Routines.java:1250)
	at org.jooq.test.OracleTest.testOracleJSR310TypesOnUDTs(OracleTest.java:4060)
Caused by: java.sql.SQLException: Error while writing value at JDBC bind index: 1
	at org.jooq.impl.DefaultBindContext.bindValue0(DefaultBindContext.java:67)
	at org.jooq.impl.AbstractBindContext.bindValue(AbstractBindContext.java:124)
	... 45 more
Caused by: org.jooq.tools.reflect.ReflectException: java.lang.reflect.InvocationTargetException
	at org.jooq.tools.reflect.Reflect.on(Reflect.java:661)
	at org.jooq.tools.reflect.Reflect.call(Reflect.java:367)
	at org.jooq.impl.DefaultBinding$DefaultArrayRecordBinding.createOracleARRAY(DefaultBinding.java:1241)
	at org.jooq.impl.DefaultBinding$DefaultArrayRecordBinding.set0(DefaultBinding.java:1163)
	at org.jooq.impl.DefaultBinding$DefaultArrayRecordBinding.set0(DefaultBinding.java:1)
	at org.jooq.impl.DefaultBinding$AbstractBinding.set(DefaultBinding.java:749)
	at org.jooq.impl.UDTRecordImpl.writeSQL(UDTRecordImpl.java:131)
	at oracle.sql.STRUCT.toSTRUCT(STRUCT.java:540)
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectCritical(OraclePreparedStatement.java:9442)
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectInternal(OraclePreparedStatement.java:8946)
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectInternal(OraclePreparedStatement.java:9766)
	at oracle.jdbc.driver.OracleCallableStatement.setObject(OracleCallableStatement.java:6068)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.setObject(OraclePreparedStatementWrapper.java:241)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.setObject(DefaultPreparedStatement.java:204)
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.set0(DefaultBinding.java:2657)
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.set0(DefaultBinding.java:1)
	at org.jooq.impl.DefaultBinding$AbstractBinding.set(DefaultBinding.java:739)
	at org.jooq.impl.DefaultBindContext.bindValue0(DefaultBindContext.java:62)
	... 46 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jooq.tools.reflect.Reflect.on(Reflect.java:657)
	... 63 more
Caused by: java.sql.SQLException: Konvertierung zu interner Darstellung nicht erfolgreich: 2017-01-01
	at oracle.jdbc.oracore.OracleTypeDATE.toDatum(OracleTypeDATE.java:71)
	at oracle.jdbc.oracore.OracleType.toDatumArray(OracleType.java:176)
	at oracle.jdbc.oracore.OracleTypeDATE.toDatumArray(OracleTypeDATE.java:108)
	at oracle.sql.ArrayDescriptor.toOracleArray(ArrayDescriptor.java:1321)
	at oracle.sql.ARRAY.<init>(ARRAY.java:140)
	at oracle.jdbc.driver.PhysicalConnection.createARRAY(PhysicalConnection.java:10278)
	... 68 more
```

**Comment by lukaseder** [comment]
Reopened this for a permanent fix. The quick fix isn't thorough. There are other conversions that might not work correctly.

**Comment by lukaseder** [comment]
Thorough fix implemented for jOOQ 3.11 in the context of #6630, #7400, #7404, #7405, and others
